### DISCUSSION TITLE
How to append an image in OnRecordBeforeCreateRequest hook?

### DISCUSSION THREAD

**Discussion by ezdevapps** [pocketbase/pocketbase#3315] [Q&A] [answered]
My formdata has an image (multiple) field, intercepted in the above hook. In this hook I want to take the first image and pass it to another external api for cleansing and append the cleansed image to the formdata before it is saved. Effectively in this hook the formdata is enriched with an additional cleansed image. Can you please suggest how can this be achieved? I prefer to use go lang.
I am able to get the first image using imgs := e.Record.Get("images")[0] but not sure about how to extract the content and pass on to another api(http) and get the blob back and append to the imgs. Appreciate your help!

**Comment by ganigeorgiev** [comment]
What do you mean by "cleansing"?
The uploaded image mimetype and file size are validated server-side and users usually don't have to do anything in addition.

Unfortunately currently the exposed files management internals are a little verbose and not very user friendly if you still need to perform some "post processing".
The main thing to note is that `e.Record` **holds only the filename(s)** (that is what we store in the database) and the files are stored separately either on the disk or in S3.

There are multiple ways to access and modify the uploaded file, but one possible approach I guess could be to use the `e.UploadedFiles` field. Here is a pseudo-code (_I haven't tested it_):

```go
app.OnRecordBeforeCreateRequest("yourCollection").Add(func(e *core.RecordCreateEvent) error {
	files, _ := e.UploadedFiles["yourFileField"]
	for i, f := range files {
		// https://pkg.go.dev/github.com/pocketbase/pocketbase/tools/filesystem#File
		// the following filesystem.File fields are available:
		// Name         string
		// OriginalName string
		// Size         int64
		// Reader       filesystem.FileReader

		// do something with the file content ...
		err := func () error {
			r, err := f.Reader.Open()
			if err != nil {
				return err
			}
			defer r.Close() 

			// ...

			return nil
		}()

		// after you modified the file, you can replace it "in place"
		// 
		// note that if you change the name, you must also replace the name in the e.Record field
		files[i], err = filesystem.NewFileFromBytes(newBytes, f.OriginalName)
	}

	return nil
})
```

**Reply by ezdevapps** [reply]
Thanks for a quick response. I am uploading an image, which is a photo with some unwanted background. To remove the background(which I referred to as "cleansing"), I need to invoke another api over http by passing the blob and the service returns the modifled blob which I want to store along with the original image. In the above code how can I retrieve the byte array (or blob) after f.Reader.Open() ?

**Reply by ganigeorgiev** [reply, marked as answer]
> In the above code how can I retrieve the byte array (or blob) after f.Reader.Open() ?

`Open()` returns a reader, meaning that you can use [`io.Copy`](https://pkg.go.dev/io#Copy) or similar to copy the content into a writer, here is an example with `bytes.Buffer`:

```go
b := new(bytes.Buffer)
_, err := io.Copy(b, r);
```

**Reply by ezdevapps** [reply]
Awesome! That worked! 
I am new to Go. I am inspired by pocketbase to learn Go. Excellent work of abstraction, I am absolutely impressed by the overall DX. Everything was very clear and I didn't face any sort of hiccups in getting started with using it. Appreciate your hard work and making it available to the dev community!
